File: scripts/generate-graph.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DOCS_DIR = path.resolve(__dirname, '../docs');
const OUTPUT_FILE = path.join(DOCS_DIR, 'graph.json');

const EDGE_TYPES = [
    "implements",
    "measured_by",
    "measures",
    "depends_on",
    "aligned_with"
];

function parseFrontMatter(content) {
    // Strip BOM if present
    content = content.replace(/^\uFEFF/, '');

    // Handle both LF and CRLF, and allow trailing spaces after ---
    const match = content.match(/^---\s*\r?\n([\s\S]*?)\r?\n---/);
    if (!match) {
        // Try to verify if it looks like it SHOULD have frontmatter but failed
        if (content.trim().startsWith('---')) {
            // console.log("Starts with --- but regex failed. Content sample:", content.slice(0, 50));
            console.log("Starts with --- but regex failed.");
            console.log("First 50 chars (hex):", Buffer.from(content.slice(0, 50)).toString('hex'));
            console.log("First 50 chars (text):", JSON.stringify(content.slice(0, 50)));
        }
        return null;
    } else {
        // console.log("Matched frontmatter!");
    }

    const yamlContent = match[1];
    const data = {};

    const lines = yamlContent.split('\n');
    for (const line of lines) {
        if (!line.trim() || line.startsWith('#')) continue;

        const parts = line.split(':');
        if (parts.length < 2) continue;

        const key = parts[0].trim();
        let value = parts.slice(1).join(':').trim();

        // Remove quotes if present
        if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
            value = value.slice(1, -1);
        }

        // Handle arrays
        if (value.startsWith('[') && value.endsWith(']')) {
            const arrayContent = value.slice(1, -1);
            if (!arrayContent.trim()) {
                data[key] = [];
            } else {
                data[key] = arrayContent.split(',').map(item => {
                    item = item.trim();
                    if ((item.startsWith('"') && item.endsWith('"')) || (item.startsWith("'") && item.endsWith("'"))) {
                        return item.slice(1, -1);
                    }
                    return item;
                });
            }
        } else {
            data[key] = value;
        }
    }

    return data;
}

function getAllFiles(dirPath, arrayOfFiles) {
    const files = fs.readdirSync(dirPath);

    arrayOfFiles = arrayOfFiles || [];

    files.forEach(function (file) {
        if (file.startsWith('_') || file.startsWith('.')) return;

        const fullPath = path.join(dirPath, file);
        if (fs.statSync(fullPath).isDirectory()) {
            arrayOfFiles = getAllFiles(fullPath, arrayOfFiles);
        } else if (file.endsWith('.md')) {
            arrayOfFiles.push(fullPath);
        }
    });

    return arrayOfFiles;
}

function generateGraph() {
    console.log(`Scanning docs directory: ${DOCS_DIR}`);
    if (!fs.existsSync(DOCS_DIR)) {
        console.error(`Docs directory not found: ${DOCS_DIR}`);
        process.exit(1);
    }

    const files = getAllFiles(DOCS_DIR);
    console.log(`Found ${files.length} markdown files.`);

    const nodes = [];
    const links = [];

    // First pass: nodes
    files.forEach(file => {
        try {
            const content = fs.readFileSync(file, 'utf8');
            const data = parseFrontMatter(content);
            if (data && data.id) {
                // Determine relative path based on execution context context
                const relativePath = path.relative(DOCS_DIR, file).replace(/\\/g, '/');
                data.source_file = relativePath;
                nodes.push(data);
            }
        } catch (err) {
            console.error(`Error parsing file ${file}:`, err.message);
        }
    });

    const nodeIds = new Set(nodes.map(n => n.id));

    // Second pass: links
    nodes.forEach(node => {
        EDGE_TYPES.forEach(edgeType => {
            if (node[edgeType]) {
                const targets = Array.isArray(node[edgeType]) ? node[edgeType] : [node[edgeType]];

                targets.forEach(targetId => {
                    if (!targetId) return;

                    links.push({
                        source: node.id,
                        target: targetId,
                        type: edgeType,
                        source_kind: node.type,
                        exists_in_docs: nodeIds.has(targetId)
                    });
                });
            }
        });
    });

    const graph = {
        nodes,
        links,
        metadata: {
            total_nodes: nodes.length,
            total_links: links.length,
            version: "1.0",
            generated_at: new Date().toISOString()
        }
    };

    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(graph, null, 2), 'utf8');
    console.log(`Graph generated at ${OUTPUT_FILE}`);
    console.log(`Stats: ${nodes.length} nodes, ${links.length} links`);
}

generateGraph();

File: scripts/generate_graph.py
import os
import yaml
import re
import json

DOCS_DIR = r"f:\RAI_EP\docs"
OUTPUT_FILE = os.path.join(DOCS_DIR, "graph.json")

EDGE_TYPES = [
    "implements",
    "measured_by",
    "measures",
    "depends_on",
    "aligned_with"
]

def parse_md_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    match = re.search(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return None
    
    try:
        data = yaml.safe_load(match.group(1))
        # Add relative path for link back
        data['source_file'] = os.path.relpath(file_path, DOCS_DIR)
        return data
    except Exception:
        return None

def generate_graph():
    nodes = []
    links = []
    
    # First pass: Collect all nodes
    for root, dirs, files in os.walk(DOCS_DIR):
        for file in files:
            if file.endswith(".md") and not file.startswith("_"):
                file_path = os.path.join(root, file)
                node_data = parse_md_file(file_path)
                if node_data and 'id' in node_data:
                    nodes.append(node_data)
    
    # Second pass: Create links
    node_ids = {n['id'] for n in nodes}
    
    for node in nodes:
        source_id = node['id']
        for edge_type in EDGE_TYPES:
            if edge_type in node:
                targets = node[edge_type]
                if isinstance(targets, str):
                    targets = [targets]
                
                for target_id in targets:
                    # Semantic direction: Source -> Target
                    links.append({
                        "source": source_id,
                        "target": target_id,
                        "type": edge_type,
                        "source_kind": node.get('type'),
                        "exists_in_docs": target_id in node_ids
                    })

    graph = {
        "nodes": nodes,
        "links": links,
        "metadata": {
            "total_nodes": len(nodes),
            "total_links": len(links),
            "version": "1.0"
        }
    }
    
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(graph, f, indent=2, ensure_ascii=False)
    
    print(f"Graph generated: {len(nodes)} nodes, {len(links)} links.")

if __name__ == "__main__":
    generate_graph()
